### PR TITLE
Add warning to additional keys in config

### LINKED ISSUE #965: warn on unknown keys at the top level
We're now warning on unknown keys at the hook level, this should also happen at the top level

### PR DESCRIPTION
**Pull request by yoavcaspi**
Resolves #965

### CODE REVIEW

File: pre_commit/clientlib.py [hunk 1]
```diff
@@ -144,6 +147,16 @@ def _entry(modname):
     )
 
 
+def warn_on_unknown_keys_at_top_level(extra, orig_keys):
+    logger.warning(
+        'Your pre-commit-config contain these extra keys: {}. '
+        'while the only valid keys are: {}.'.format(
```
Comment (asottile) [inline pre_commit/clientlib.py]: let's keep this short and sweet, how about
```python
'Unexpected key(s): {}'.format(', '.join(sorted(extra)))
```

File: pre_commit/clientlib.py [hunk 2]
```diff
@@ -222,6 +235,10 @@ def _entry(modname):
     ),
 
     MigrateShaToRev(),
+    cfgv.WarnAdditionalKeys(
+        {'repo', 'rev', 'hooks'},
```
Comment (asottile) [inline pre_commit/clientlib.py]: let's use a tuple here instead

File: pre_commit/clientlib.py [hunk 3]
```diff
@@ -222,6 +235,10 @@ def _entry(modname):
     ),
 
     MigrateShaToRev(),
+    cfgv.WarnAdditionalKeys(
```
Comment (asottile) [inline pre_commit/clientlib.py]: can you add a quick test to make sure this doesn't drift from the `CONFIG_REPO_DICT` definition?

Basically will need to iterate through the things in that `Map`'s `.items` and check the keys against the keys passed in here

File: pre_commit/clientlib.py (hunk 2 above)
Comment (yoavcaspi) [reply pre_commit/clientlib.py]: why tuple?

File: pre_commit/clientlib.py (hunk 1 above)
Comment (yoavcaspi) [reply pre_commit/clientlib.py]: are you sure it will be clear to the user what is the problem here?

File: pre_commit/clientlib.py (hunk 2 above)
Comment (asottile) [reply pre_commit/clientlib.py]: immutability, WarnAdditionalKeys takes any `Sequence` and will convert it to a set when comparing the keys

File: pre_commit/clientlib.py (hunk 1 above)
Comment (asottile) [reply pre_commit/clientlib.py]: hmm then maybe `Unexpected config key(s): ...`?  The other message has context of location but it looks like we didn't think about that when adding `WarnAdditionalKeys` 🤔 https://github.com/pre-commit/pre-commit/blob/1c7ac67fba963b7a8cda852ed04bceb8a1c37063/pre_commit/repository.py#L108

File: pre_commit/clientlib.py (hunk 3 above)
Comment (yoavcaspi) [reply pre_commit/clientlib.py]: I'm not following. can you elaborate?

From what I understand you wish to add a test to validate that the new attribute does not change the definition of CONFIG_REPO_DICT and just added a kind of decorator. but I'm not sure how to do it.

Comment (asottile) [reply pre_commit/clientlib.py]: ~something like this

```diff
--- a/pre_commit/clientlib.py
+++ b/pre_commit/clientlib.py
@@ -109,6 +109,8 @@ META = 'meta'
 
 
 class MigrateShaToRev(object):
+    key = 'rev'
+
     @staticmethod
     def _cond(key):
         return cfgv.Conditional(
diff --git a/tests/clientlib_test.py b/tests/clientlib_test.py
index cace0f3..13b42a5 100644
--- a/tests/clientlib_test.py
+++ b/tests/clientlib_test.py
@@ -305,3 +305,12 @@ def test_minimum_pre_commit_version_failing():
 def test_minimum_pre_commit_version_passing():
     cfg = {'repos': [], 'minimum_pre_commit_version': '0'}
     cfgv.validate(cfg, CONFIG_SCHEMA)
+
+
+@pytest.mark.parametrize('schema', (CONFIG_SCHEMA, CONFIG_REPO_DICT))
+def test_warn_additional(schema):
+    allowed_keys = {item.key for item in schema.items if hasattr(item, 'key')}
+    warn_additional, = [
+        x for x in schema.items if isinstance(x, cfgv.WarnAdditionalKeys)
+    ]
+    assert allowed_keys == set(warn_additional.keys)
```

Comment (yoavcaspi) [reply pre_commit/clientlib.py]: Done.
This test helped me find some more issues with the code. Thanks :-)

**Comment by yoavcaspi** [comment]
Let me know if there are other issues.

File: pre_commit/clientlib.py [hunk 4]
```diff
@@ -144,6 +149,14 @@ def _entry(modname):
     )
 
 
+def warn_on_unknown_keys_at_top_level(extra, orig_keys):
```
Comment (asottile) [inline pre_commit/clientlib.py]: probably just `warn_unknown_keys` now that it's being used in other places

Comment (yoavcaspi) [reply pre_commit/clientlib.py]: cool. thanks.

**Comment by asottile** [comment]
Thanks again!  And thanks to @mitzkia for the assist via `cfgv`!

**Pull request merged**